First implementation (#1)
pandosearch/traffic_drain_plug: README.md

# TrafficDrain Plug

This is the codification of the example `Plug` implementation as documented in [Botsquad's
GracefulStop Elixir Package](https://github.com/botsquad/graceful_stop)

The `Plug` provides a special endpoint named `__traffic` that will respond with an HTTP 200 OK

> statuscode whilst the Phoenix Application is ready to handle requests. Whenever the beam process

Reviewer: ```suggestion
statuscode whilst the [Phoenix Application](https://www.phoenixframework.org/) is ready to handle requests. Whenever the beam process
```

Author: Als je met een Plug aan de slag gaat weet je wel waar je Phoenix kunt vinden.

> statuscode whilst the Phoenix Application is ready to handle requests. Whenever the beam process

Reviewer 2: ```suggestion
status code as long as the Phoenix Application is ready to handle requests. Whenever the BEAM process
```

has received a `SIGTERM` signal by k8s in order to go down, this `__traffic` endpoint will start
to respond with an HTTP 503 Service Unavailable statuscode.

> By combining this behaviour with a kubernetes [readinessProbe](https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/#define-readiness-probes)

Reviewer 2: ```suggestion
By combining this behaviour with a Kubernetes [readinessProbe](https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/#define-readiness-probes)
```

> you will have a setup that is able to provide seamless upgrades of Phoenix Web Apps on Kubernetes.

Reviewer 2: ```suggestion
you will have a setup that is able to provide seamless upgrades of Phoenix web apps on Kubernetes.
```

## Credits

> The original idea for capturing the `SIGTERM` signal and act accordingly is documented in [this blog

Reviewer 2: ```suggestion
The original idea for capturing the `SIGTERM` signal and acting accordingly is documented in [this blog
```

post](https://medium.com/@ellispritchard/graceful-shutdown-on-kubernetes-with-signals-erlang-otp-20-a22325e8ae98).

> It has an accompanying [repo on github](https://github.com/Financial-Times/k8s_traffic_plug).

Reviewer 2: ```suggestion
It has an accompanying [repo on GitHub](https://github.com/Financial-Times/k8s_traffic_plug).
```

[Botsquad](https://www.botsquad.com) has implemented its [own version of this idea](https://github.com/botsquad/graceful_stop)
which is more flexible than its predecessor, but does not contain a `Plug`. However in the

> [README](https://github.com/botsquad/graceful_stop/blob/master/README.md) there is an example,

Reviewer 2: ```suggestion
[README](https://github.com/botsquad/graceful_stop/blob/master/README.md) there is an example
```

> written by [Vladimir Gorej](https://github.com/char0n), for a `Plug` which is the code we

Reviewer 2: ```suggestion
for a `Plug`, written by [Vladimir Gorej](https://github.com/char0n). This is the code we
```

> have used to create this `TrafficDrain Plug`.

Reviewer 2: ```suggestion
have used to create this TrafficDrain `Plug`.
```

## Installation

Add this repository to your dependencies. As we've merely codified an example we think that we are
not the right people to publish this `Plug`, so if you want to use this `Plug` you'll have to use

> a [git reference](https://hexdocs.pm/mix/Mix.Tasks.Deps.html#module-git-options-git).

Reviewer 2: ```suggestion
a [Git reference](https://hexdocs.pm/mix/Mix.Tasks.Deps.html#module-git-options-git).
```

```elixir
def deps do
  [
    ...,
    {:traffic_drain_plug, github: "pandosearch/traffic_drain_plug"}
  ]
end
```

## Usage

> Add this `Plug` in your [Phoenix Endpoint](https://hexdocs.pm/phoenix/Phoenix.Endpoint.html#content)

Reviewer 2: ```suggestion
Add this `Plug` to your [Phoenix Endpoint](https://hexdocs.pm/phoenix/Phoenix.Endpoint.html#content).
```

> before your Router Plug. You don't need any of the router functionality, so it should run

Reviewer 2: ```suggestion
As you don't want any router functionality to interfere, make sure this plug runs
```

> **before** the Phoenix Router.

Reviewer 2: ```suggestion
**before** the Phoenix Router:
```
